hdr/sqlite_modern_cpp.h

```diff
@@ -85,11 +85,19 @@ namespace sqlite {
 			return ++_inx;
 		}
 
-		sqlite3_stmt* _prepare(const std::u16string& sql) {
-			return _prepare(utility::utf16_to_utf8(sql));
+		sqlite3_stmt* _prepare(const U16STR_REF& sql) {
+			//return _prepare(utility::utf16_to_utf8(sql));
+			int hresult;
+			sqlite3_stmt* tmp = nullptr;
+			const void *remaining;
+			hresult = sqlite3_prepare16_v2(_db.get(), sql.data(), -1, &tmp, &remaining);
+			if (hresult != SQLITE_OK) errors::throw_sqlite_error(hresult, utility::utf16_to_utf8(sql.data()));
+			if (!std::all_of(static_cast<const char16_t*>(remaining), sql.data() + sql.size(), [](char16_t ch) {return std::isspace(ch); }))
+				throw errors::more_statements("Multiple semicolon separated statements are unsupported", utility::utf16_to_utf8(sql.data()));
+			return tmp;
 		}
 
-		sqlite3_stmt* _prepare(const std::string& sql) {
+		sqlite3_stmt* _prepare(const STR_REF& sql) {
```

Comment:
Can I ask what the advantage of this would be? I know that in my use case, I'm going to have a few `string_view` objects already around and wouldn't passing by const reference be marginally faster than passing by value?

Reply:
In most cases they will be inlined, then it doesn't make a difference. Otherwise the advantage is that a const reference requires some object to reference, so the `string_view` has to exists in memory(typically on the stack). But when passing by value small objects like this normally are passed in registers which does not require any usage of (slow) memory.
Most of the time the string_view will not exist in memory before(because it is a temporary which did not exists or because the optimizer keeps the values in registers), so copying `2*sizeof(void*)` between registers should be much faster than writing this into memory, copying the pointer and reading from memory again.
This does mostly apply for the Itanium ABI, so not on Windows. But output from Compiler-Explorer suggests that passing by value saves a memory access on Windows too.
Additionally the interface looks cleaner :wink:

Disclaimer: I did not run a benchmark on this so I could be wrong. In this case, I would love to hear why.

Reply:
You're probably right. I wrote some super crappy benchmarks and it seems passing by string_view was slightly faster: https://github.com/zowpowow/sqlite_modern_cpp/tree/str_view_by_value